lib/widgets/topic_list.dart

```diff
@@ -0,0 +1,333 @@
+          child: Icon(size: 18, icon, color: channelIconColor)),
+        Flexible(child: Text(
+          stream?.name ?? zulipLocalizations.unknownChannelName,
+          style: TextStyle(
+            fontSize: 20,
+            height: 30 / 20,
+            color: designVariables.title,
+          ).merge(weightVariableTextStyle(context, wght: 600)))),
+      ]);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final alignment = willCenterTitle
+      ? Alignment.center
+      : AlignmentDirectional.centerStart;
+    return SizedBox(
+      width: double.infinity,
+      child: GestureDetector(
+        behavior: HitTestBehavior.translucent,
+        onLongPress: () {
+          showChannelActionSheet(context, channelId: streamId);
+        },
+        child: Align(alignment: alignment,
+          child: _buildAppBarRow(context))));
+  }
+}
+
+class _TopicList extends StatefulWidget {
+  const _TopicList({required this.streamId});
+
+  final int streamId;
+
+  @override
+  State<_TopicList> createState() => _TopicListState();
+}
+
+class _TopicListState extends State<_TopicList> with PerAccountStoreAwareStateMixin {
+  Unreads? unreadsModel;
+  // TODO(#1499): store the results on [ChannelStore], and keep them
+  //   up-to-date by handling events
+  List<GetStreamTopicsEntry>? lastFetchedTopics;
+
+  @override
+  void onNewStore() {
+    unreadsModel?.removeListener(_modelChanged);
+    final store = PerAccountStoreWidget.of(context);
+    unreadsModel = store.unreads..addListener(_modelChanged);
+    _fetchTopics();
+  }
+
+  @override
+  void dispose() {
+    super.dispose();
+    unreadsModel?.removeListener(_modelChanged);
+  }
+
+  void _modelChanged() {
+    setState(() {
+      // The actual state lives in `unreadsModel`.
+    });
+  }
+
+  void _fetchTopics() async {
+    // Do nothing when the fetch fails; the topic-list will stay on
+    // the loading screen, until the user navigates away and back.
+    // TODO(design) show a nice error message on screen when this fails
+    final store = PerAccountStoreWidget.of(context);
+    final result = await getStreamTopics(store.connection,
+      streamId: widget.streamId,
+      allowEmptyTopicName: true);
+    if (!mounted) return;
+    setState(() {
+      lastFetchedTopics = result.topics;
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    if (lastFetchedTopics == null) {
+      return const Center(child: CircularProgressIndicator());
+    }
+
+    // TODO(design) handle the rare case when `lastFetchTopics` is empty
+
+    // This is adapted from parts of the build method on [_InboxPageState].
+    final topicItems = <_TopicItemData>[];
+    for (final GetStreamTopicsEntry(:maxId, name: topic) in lastFetchedTopics!) {
+      final unreadMessageIds =
+        unreadsModel!.streams[widget.streamId]?[topic] ?? <int>[];
+      final countInTopic = unreadMessageIds.length;
```

Comment:
Let's centralize on `Unreads.countInTopicNarrow` for this; I could imagine wanting to refine it for something like muted senders etc.

Reply:
We might still need to keep `unreadMessageIds` to compute the value of `hasMention`.
Looks like inbox page does something like that with DMs:

```dart
      final countInNarrow = unreadsModel!.countInDmNarrow(dmNarrow);
      if (countInNarrow == 0) {
        continue;
      }
      final hasMention = unreadsModel!.dms[dmNarrow]!.any(
        (messageId) => unreadsModel!.mentions.contains(messageId));
```